src/game.rs

```diff
@@ -21,90 +21,237 @@ pub enum ChessError {
+        Self {
+            history: vec![(state, Move::default())],
+            ..Default::default()
+        }
+    }
+
+    fn add_info(&mut self, next_state: BoardState, mv: Move) {
+        self.history.push((next_state, mv));
+        self.current_turn += 1
+    }
+
+    fn get_current_state(&self) -> &BoardState {
+        &self.history[self.current_turn - 1].0
+    }
+
+    fn get_info_for_turn(&self, turn: usize) -> &(BoardState, Move) {
+        &self.history[turn]
+    }
+
+    fn resume(&mut self) {
+        self.current_turn = self.history.len()
+    }
+
+    fn previous_state(&mut self) {
+        if self.current_turn > 1 {
+            self.current_turn -= 1
+        }
+    }
+
+    fn next_state(&mut self) {
+        if self.current_turn < self.history.len() {
+            self.current_turn += 1
+        }
+    }
+
+    fn initial_state(&mut self) {
+        self.current_turn = 1
+    }
+
+    fn is_ongoing(&mut self) -> bool {
+        self.current_turn == self.history.len()
+    }
+}
+
+#[derive(Clone)]
 pub struct Game {
-    state: BoardState,
     valid_moves: HashSet<Move>,
     status: GameStatus,
+    history: History,
+}
+
+impl Default for Game {
+    fn default() -> Self {
+        Game::with_history(History::with_state(BoardState::default()))
+    }
 }
 
 impl Game {
     pub fn new() -> Self {
-        let mut game = Self::default();
+        Game::default()
+    }
+
+    pub fn with_history(history: History) -> Self {
+        let mut game = Self {
+            history,
+            ..Default::default()
+        };
         game.add_moves();
         game
     }
 
+    pub fn with_state(state: BoardState) -> Self {
+        Game::with_history(History::with_state(state))
+    }
+
     pub fn get_piece(&self, position: &Position) -> Option<Piece> {
-        self.state.get_piece(position)
+        self.history.get_current_state().get_piece(position)
     }
 
-    pub fn get_board_state(&self) -> &BoardState {
-        &self.state
+    fn get_current_state(&self) -> &BoardState {
+        self.history.get_current_state()
+    }
+
+    fn get_current_player(&self) -> Player {
+        self.get_current_state().player
+    }
+
+    fn is_piece_some(&self, at: &Position) -> ChessResult {
+        self.get_current_state().is_piece_some(at)
+    }
+
+    pub fn has_piece(&self, position: &Position) -> bool {
+        self.history.get_current_state().has_piece(position)
+    }
+
+    fn piece_can_snipe(&self, at: &Position) -> bool {
+        self.get_piece(at).unwrap().can_snipe()
+    }
+
+    fn get_info_for_turn(&self, turn: usize) -> &(BoardState, Move) {
+        self.history.get_info_for_turn(turn)
+    }
+
+    fn has_castling_right(&self, right: CastlingRightsKind) -> bool {
+        self.get_current_state()
+            .castling_rights
+            .has_castling_right(right)
+    }
+
+    pub fn go_back_a_turn(&mut self) {
+        self.status.update(GameStatus::Replay);
```

Comment:
So I think we should only update the status in `update_status` so you don't have to think to hard in the future. You'd just call `update_status` at the end of these functions, or even just `update` if you wanted the `valid_moves` to match the state that's being replayed.